packages/teuchos/parameterlist/src/Teuchos_ParameterList.hpp

```diff
@@ -951,6 +962,15 @@ ParameterList& ParameterList::set(
   return *this;
 }
 
+template<typename T, typename S, typename>
+inline
+ParameterList& ParameterList::set(
+  std::string const& name_in, S&& value_in, std::string const& docString_in,
+  RCP<const ParameterEntryValidator> const& validator_in
+  )
+{
+  return set<S>(name_in, std::forward<S>(value_in), docString_in, validator_in);
```

Comment:
I am confused here.  Should this not be stored as `T`, the first template parameter?  Looking at:

* https://github.com/uliegecsm/Trilinos/blob/a134ec356537b3a86fc53044a2e3c7e0c62cc5fe/packages/teuchos/parameterlist/src/Teuchos_ParameterList.hpp#L956

it seems like this call would story this as the 2nd template type `S`?  Is this to allow `S` to be `const std::string&` while `T` is `std::string`, like in the below unit tests?

Reply:
Indeed, the goal is to allow use cases like the following one from the tests:

```
std::string my_str_3{"my text 3"};
pl.set<std::string>("my string 3", my_str_3);
```

Such use cases caused the failures in the pipeline with the original version of the PR. The issue with such use cases is that the template parameter is specified at the call site (typically without reference), and so it can no longer be deduced (in an appropriate reference category as needed for the perfect forwarding). In other words, if we have only the perfect forwarding version of the templated set method, then specifying the template parameter at the call site in this way forces that the parameter value be passed by rvalue reference. But, this is not possible here because the parameter value is a named variable at the call site, so it cannot be moved. This is why the fallback is needed.

If we used `return set<T>(...` in the fallback on line 972, we'd recreate at that call site the issue.

I think it would work to write just `return set(...` without specifying the template parameter. But then, the compiler can be expected to deduce precisely `S` when it would go through deducing the template parameters for the call on line 923. And so, the idea is to write `return set<S>` here to avoid this work for the compiler.

Reply:
It looks like the auto merge label led to the pr to be merged. I hope you are fine with this. If you should want to make changes, I would be happy to.

Reply:
Thanks for the explanation.  There are several applications at SNL that test against Trilinos within 24 hours and we will hear about a problem if one exists.  Thanks!